numpyro/infer/autoguide.py

```diff
@@ -864,6 +865,314 @@ def _single_sample(_rng_key):
+                event_dim=0,
+            )
+            eta_coeff = numpyro.param("{}_eta_coeff".format(self.prefix), jnp.zeros(N), event_dim=0)
+
+            gamma = numpyro.param(
+                "{}_gamma".format(self.prefix),
+                jnp.ones(N) * 0.9,
+                constraint=constraints.interval(0, 1),
+                event_dim=0,
+            )
+            betas = numpyro.param(
+                "{}_beta_increments".format(self.prefix),
+                jnp.ones((N, K)),
+                constraint=constraints.positive,
+                event_dim=1,
+            )
+            betas = jnp.cumsum(betas, axis=-1)
+            betas = betas / betas[..., -1:]
+
+            mass_matrix = numpyro.param(
+                "{}_mass_matrix".format(self.prefix),
+                jnp.ones((N, D)),
+                constraint=constraints.positive,
+                event_dim=1,
+            )
+            inv_mass_matrix = 0.5 / mass_matrix
+            assert inv_mass_matrix.shape == (subsample_size, D)
+            z_0_loc_init = jnp.zeros((N, D))
+            z_0_loc = numpyro.param("{}_z_0_loc".format(self.prefix), z_0_loc_init, event_dim=1)
+            z_0_scale_init = jnp.ones((N, D)) * self.init_scale
+            z_0_scale = numpyro.param(
+                "{}_z_0_scale".format(self.prefix),
+                z_0_scale_init,
+                constraint=constraints.positive,
+                event_dim=1,
+            )
+            base_z_dist = dist.Normal(z_0_loc, z_0_scale).to_event(1)
+            assert base_z_dist.shape() == (subsample_size, D)
+            z_0 = numpyro.sample("{}_z_0".format(self.prefix), base_z_dist, infer={"is_auxiliary": True})
+            base_z_dist_log_prob = lambda x: base_z_dist.log_prob(x).sum()
+
+            momentum_dist = dist.Normal(0, mass_matrix).to_event(1)
+            eps = numpyro.sample(
+                "{}_momentum".format(self.prefix),
+                dist.Normal(0, mass_matrix[..., None])
+                .expand([subsample_size, D, K])
+                .to_event(2)
+                .mask(False),
+                infer={"is_auxiliary": True},
+            )
+
+            local_log_density = make_local_log_density(global_output,
+                _subsample_idx={plate_name: idx})
+
+            def scan_body(carry, eps_beta):
+                eps, beta = eps_beta
+                eta = eta0 + eta_coeff * beta
+                eta = jnp.clip(eta, a_min=0.0, a_max=self.eta_max)
+                assert eps.shape == (subsample_size, D)
+                assert eta.shape == beta.shape == (subsample_size,)
+                z_prev, v_prev, log_factor = carry
+                z_half = z_prev + v_prev * eta[:, None] * inv_mass_matrix
+                q_grad = (1.0 - beta[:, None]) * grad(base_z_dist_log_prob)(z_half)
+                # Note: rescale log_density to be order 1.
+                # Q: is it true that local_log_density is scaled by (N / subsample_size) because of the plate?
```

Comment:
is it true that local_log_density is scaled by (N / subsample_size) because of the plate?

Reply:
if this is true we need the multiplicative factor in 

```
p_grad = beta[:, None] * (subsample_size / N) * grad(local_log_density)(z_half)
```
because `p_grad` and `q_grad` need to have the same scale. `q_grad` comes from `base_z_dist_log_prob` so i guess it should be unscaled

Reply:
You  are right, just double check that we need such factor here.